controllers/workflow_controller_test.go

```diff
@@ -0,0 +1,195 @@
+			driverStatus,
+			statusWrongDriver,
+			statusWrongWatchState,
+			statusCompleted,
+			statusError,
+		}
+
+		ct := metav1.NowMicro()
+		expectedDriverStatuses := []dwsv1alpha1.WorkflowDriverStatus{
+			{
+				DriverID:     DRIVERID,
+				DWDIndex:     0,
+				WatchState:   dwsv1alpha1.StateProposal,
+				Status:       dwsv1alpha1.StatusCompleted,
+				Completed:    true,
+				CompleteTime: &ct,
+			},
+			statusWrongDriver,
+			statusWrongWatchState,
+			statusCompleted,
+			statusError,
+		}
+
+		Expect(k8sClient.Create(context.TODO(), wf)).To(Succeed())
+
+		Eventually(func(g Gomega) []dwsv1alpha1.WorkflowDriverStatus {
+			returnWf := &dwsv1alpha1.Workflow{}
+			g.Expect(k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(wf), returnWf)).To(Succeed())
+			driverStatus = returnWf.Status.Drivers[0]
+			return returnWf.Status.Drivers
+		}).Should(BeComparableTo(expectedDriverStatuses,
+			cmp.Comparer(func(x, y *metav1.MicroTime) bool {
+				// Don't compare times, just check for Nil
+				bothSet := !(reflect.ValueOf(x).IsNil() || reflect.ValueOf(y).IsNil())
+				bothUnset := reflect.ValueOf(x).IsNil() && reflect.ValueOf(y).IsNil()
+				return bothSet || bothUnset
+			}),
+		))
+	})
+
+	It("Can set Workflow driver errors", func() {
+		action := "error"
+		message := "Test_error_message"
+		wf.Spec.DWDirectives = []string{
+			fmt.Sprintf("#DW STATUS action=%s message=%s", action, message),
+		}
+
+		driverStatus := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+		}
+		wf.Status.Drivers = []dwsv1alpha1.WorkflowDriverStatus{driverStatus}
+
+		expectedDriverStatus := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+			Status:     dwsv1alpha1.StatusError,
+			Error:      "Test error message",
+		}
+
+		Expect(k8sClient.Create(context.TODO(), wf)).To(Succeed())
+
+		Eventually(func(g Gomega) dwsv1alpha1.WorkflowDriverStatus {
+			returnWf := &dwsv1alpha1.Workflow{}
+			g.Expect(k8sClient.Get(context.TODO(), client.ObjectKeyFromObject(wf), returnWf)).To(Succeed())
+			return returnWf.Status.Drivers[0]
+		}).Should(Equal(expectedDriverStatus))
+	})
+
+	It("Can No-op Workflow driver statuses", func() {
+		action := "wait"
+		wf.Spec.DWDirectives = []string{
+			fmt.Sprintf("#DW STATUS action=%s", action),
+		}
+
+		driverStatus := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+		}
+		wf.Status.Drivers = []dwsv1alpha1.WorkflowDriverStatus{driverStatus}
+
+		// The driver status should be unchanged
+		expectedDriverStatus := dwsv1alpha1.WorkflowDriverStatus{
+			DriverID:   DRIVERID,
+			DWDIndex:   0,
+			WatchState: dwsv1alpha1.StateProposal,
+		}
+
+		Expect(k8sClient.Create(context.TODO(), wf)).To(Succeed())
+
+		Eventually(func(g Gomega) dwsv1alpha1.WorkflowDriverStatus {
```

Comment:
This never loops, does it?

My question when I read this is, "for how long?"   Maybe instead of Eventually(), you want to use Consistently().
https://onsi.github.io/gomega/#consistently

Reply:
"Eventually" seems to work exactly like I need. It's also the function used in the NNF tests.